File: server/internal/mcpserver/projections_budget.go
package mcpserver

import (
	"tallyo/internal/graph/model"
	"tallyo/internal/money"
	u "tallyo/internal/utils"
)

type leanBudget struct {
	ID           string      `json:"id"`
	Month        string      `json:"month"`
	CategoryID   string      `json:"categoryId"`
	CategoryName string      `json:"categoryName"`
	Amount       money.Cents `json:"amount"`
}

func mapBudget(budget *model.Budget) leanBudget {
	return leanBudget{ID: budget.ID.EncodedString(), Month: budget.Month, CategoryID: budget.Category.ID.EncodedString(), CategoryName: budget.Category.Name, Amount: budget.Amount}
}

type leanBudgetLine struct {
	ID           *string     `json:"id,omitempty"`
	CategoryID   string      `json:"categoryId"`
	CategoryName string      `json:"categoryName"`
	Budgeted     money.Cents `json:"budgeted"`
	Actual       money.Cents `json:"actual"`
	Remaining    money.Cents `json:"remaining"`
}

type leanBudgetSection struct {
	Label     string           `json:"label"`
	GroupID   string           `json:"groupId"`
	GroupName string           `json:"groupName"`
	Budgeted  money.Cents      `json:"budgeted"`
	Actual    money.Cents      `json:"actual"`
	Remaining money.Cents      `json:"remaining"`
	Lines     []leanBudgetLine `json:"lines"`
}

type leanBudgetReport struct {
	Month             string              `json:"month"`
	ExpensesBudgeted  money.Cents         `json:"expensesBudgeted"`
	ExpensesActual    money.Cents         `json:"expensesActual"`
	IncomeBudgeted    money.Cents         `json:"incomeBudgeted"`
	IncomeActual      money.Cents         `json:"incomeActual"`
	RemainingBudgeted money.Cents         `json:"remainingBudgeted"`
	RemainingActual   money.Cents         `json:"remainingActual"`
	Sections          []leanBudgetSection `json:"sections"`
}

func mapBudgetReport(report *model.BudgetReport) *leanBudgetReport {
	toLine := func(l *model.BudgetLine) leanBudgetLine {
		ref := mapCategoryRef(l.Category)
		var id *string
		if l.ID != nil {
			encoded := l.ID.EncodedString()
			id = &encoded
		}
		return leanBudgetLine{
			ID:           id,
			CategoryID:   ref.ID,
			CategoryName: ref.Name,
			Budgeted:     l.Budgeted,
			Actual:       l.Actual,
			Remaining:    l.Remaining,
		}
	}
	toSection := func(s *model.BudgetSection) leanBudgetSection {
		return leanBudgetSection{
			Label:     s.Label,
			GroupID:   s.Group.ID.EncodedString(),
			GroupName: s.Group.Name,
			Budgeted:  s.Budgeted,
			Actual:    s.Actual,
			Remaining: s.Remaining,
			Lines:     u.Map(s.Lines, toLine),
		}
	}
	return &leanBudgetReport{
		Month:             report.Month,
		ExpensesBudgeted:  report.ExpensesBudgeted,
		ExpensesActual:    report.ExpensesActual,
		IncomeBudgeted:    report.IncomeBudgeted,
		IncomeActual:      report.IncomeActual,
		RemainingBudgeted: report.RemainingBudgeted,
		RemainingActual:   report.RemainingActual,
		Sections:          u.Map(report.Sections, toSection),
	}
}

File: server-rs/src/mcpserver/projections_budget.rs
use serde::Serialize;

use super::projections::{encode, map_category_ref};
use crate::{
    ids::GlobalIdType,
    money::Cents,
    schema::{Budget, BudgetLine, BudgetReport, BudgetSection},
};

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanBudget {
    pub(super) id: String,
    pub(super) month: String,
    pub(super) category_id: String,
    pub(super) category_name: String,
    pub(super) amount: Cents,
}

#[derive(Debug, PartialEq, Serialize)]
pub(super) struct LeanBudgetPayload {
    pub(super) budget: LeanBudget,
}

pub(super) fn map_budget(budget: Budget) -> LeanBudget {
    LeanBudget {
        id: encode(GlobalIdType::Budget, budget.id),
        month: budget.month,
        category_id: encode(GlobalIdType::Category, budget.category.id),
        category_name: budget.category.name,
        amount: budget.amount,
    }
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanBudgetLine {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) id: Option<String>,
    pub(super) category_id: String,
    pub(super) category_name: String,
    pub(super) budgeted: Cents,
    pub(super) actual: Cents,
    pub(super) remaining: Cents,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanBudgetSection {
    pub(super) label: String,
    pub(super) group_id: String,
    pub(super) group_name: String,
    pub(super) budgeted: Cents,
    pub(super) actual: Cents,
    pub(super) remaining: Cents,
    pub(super) lines: Vec<LeanBudgetLine>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanBudgetReport {
    pub(super) month: String,
    pub(super) expenses_budgeted: Cents,
    pub(super) expenses_actual: Cents,
    pub(super) income_budgeted: Cents,
    pub(super) income_actual: Cents,
    pub(super) remaining_budgeted: Cents,
    pub(super) remaining_actual: Cents,
    pub(super) sections: Vec<LeanBudgetSection>,
}

pub(super) fn map_budget_report(report: BudgetReport) -> LeanBudgetReport {
    let to_line = |line: BudgetLine| {
        let category = map_category_ref(&line.category);
        LeanBudgetLine {
            id: line.id.map(|id| id.to_string()),
            category_id: category.id,
            category_name: category.name,
            budgeted: line.budgeted,
            actual: line.actual,
            remaining: line.remaining,
        }
    };
    let to_section = |section: BudgetSection| LeanBudgetSection {
        label: section.label,
        group_id: encode(GlobalIdType::CategoryGroup, section.group.id),
        group_name: section.group.name,
        budgeted: section.budgeted,
        actual: section.actual,
        remaining: section.remaining,
        lines: section.lines.into_iter().map(to_line).collect(),
    };
    LeanBudgetReport {
        month: report.month,
        expenses_budgeted: report.expenses_budgeted,
        expenses_actual: report.expenses_actual,
        income_budgeted: report.income_budgeted,
        income_actual: report.income_actual,
        remaining_budgeted: report.remaining_budgeted,
        remaining_actual: report.remaining_actual,
        sections: report.sections.into_iter().map(to_section).collect(),
    }
}
